core/store/src/trie/mem/resharding.rs

```diff
@@ -59,110 +87,166 @@ impl<'a, M: ArenaMemory> MemTrieUpdate<'a, M> {
-                return;
+                self.generic_place_node(node_id, GenericUpdatedTrieNodeWithSize::empty());
+                return Ok(());
             }
-            UpdatedMemTrieNode::Leaf { extension, value } => {
+            GenericUpdatedTrieNode::Leaf { extension, value } => {
                 let full_key_nibbles =
                     [key_nibbles, NibbleSlice::from_encoded(&extension).0.iter().collect_vec()]
                         .concat();
                 if !intervals_nibbles.iter().any(|interval| interval.contains(&full_key_nibbles)) {
-                    self.place_node(node_id, UpdatedMemTrieNode::Empty);
+                    self.generic_place_node(node_id, GenericUpdatedTrieNodeWithSize::empty());
                 } else {
-                    self.place_node(node_id, UpdatedMemTrieNode::Leaf { extension, value });
+                    self.generic_place_node(
+                        node_id,
+                        GenericUpdatedTrieNodeWithSize {
+                            node: GenericUpdatedTrieNode::Leaf { extension, value },
+                            memory_usage,
+                        },
+                    );
                 }
-                return;
+                return Ok(());
             }
-            UpdatedMemTrieNode::Branch { mut children, mut value } => {
+            GenericUpdatedTrieNode::Branch { mut children, mut value } => {
                 if !intervals_nibbles.iter().any(|interval| interval.contains(&key_nibbles)) {
                     value = None;
                 }
 
+                let mut memory_usage = 0;
                 for (i, child) in children.iter_mut().enumerate() {
                     let Some(old_child_id) = child.take() else {
                         continue;
                     };
 
-                    let new_child_id = self.ensure_updated(old_child_id);
+                    let new_child_id = self.generic_ensure_updated(old_child_id)?;
                     let child_key_nibbles = [key_nibbles.clone(), vec![i as u8]].concat();
-                    self.retain_multi_range_recursive(
+                    self.generic_retain_multi_range_recursive(
                         new_child_id,
                         child_key_nibbles,
                         intervals_nibbles,
-                    );
-                    if self.updated_nodes[new_child_id] == Some(UpdatedMemTrieNode::Empty) {
+                    )?;
+                    if matches!(
+                        self.generic_get_node(new_child_id).node,
+                        GenericUpdatedTrieNode::Empty
+                    ) {
                         *child = None;
                     } else {
-                        *child = Some(OldOrUpdatedNodeId::Updated(new_child_id));
+                        *child = Some(GenericNodeOrIndex::Updated(new_child_id));
+                        memory_usage += self.generic_get_node(new_child_id).memory_usage;
```

Comment:
nit: it case it matters for performance can do `self.generic_get_node(new_child_id)` once before if-else